src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(tail, value, null);
+        if (isEmpty()) {
+            head = node;
+        } else {
+            tail.next = node;
+        }
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> node = new Node<>(null, value, null);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index - " + index
+                    + " out of bounds for length: " + size);
+        }
+        if (index == 0) {
+            node.prev = null;
+            node.next = head;
+            head.prev = node;
+            head = node;
+        } else {
+            Node<T> temp = getNode(index);
+            node.prev = temp.prev;
+            node.next = temp;
+            temp.prev.next = node;
+            temp.prev = node;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).currentValue;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        T newValue = node.currentValue;
+        node.currentValue = value;
+        return newValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
```

Comment:
No need to do this check, you already do it in `getNode(index)` on the second line